ride/referral.ride

```diff
@@ -0,0 +1,197 @@
+
+func getStringOrFail(key: String) = this.getString(key).valueOrErrorMessage(key + " is not defined")
+
+let marketingContract = keyMarketingContract.getStringOrFail().addressFromStringValue()
+let gwxRewardContract = keyMarketingContract.getStringOrFail().addressFromStringValue()
+
+func keyExistsReferrerToReferral(
+  programName: String,
+  referrerAddress: String,
+  referralAddress: String
+) = ["%s%s%s%s", "existsReferrerToReferral", programName, referrerAddress, referralAddress].makeString(separator)
+func keyClaimed(
+  programName: String,
+  claimerAddress: String
+) = makeString(["%s%s%s%s", "claimed", programName, claimerAddress], separator)
+func keyUnclaimed(
+  programName: String,
+  claimerAddress: String
+) = makeString(["%s%s%s%s", "unclaimed", programName, claimerAddress], separator)
+func keyReferrer(
+  programName: String,
+  referralAddress: String
+) = makeString(["%s%s%s", "referrer", programName, referralAddress], separator)
+func keyProgramName(programName: String) = makeString(["%s%s", "programName", programName], separator)
+func keyTotalReferralCount(
+  programName: String,
+  referrerAddress: String
+) = makeString(["%s%s%s", "totalReferralCount", programName, referrerAddress], separator)
+func keyActiveReferralCount(
+  programName: String,
+  referrerAddress: String
+) = makeString(["%s%s%s", "activeReferralCount", programName, referrerAddress], separator)
+func keyProgramAllowedContract(
+  programName: String
+) = makeString(["%s%s%s", "incAllowedContract", programName], separator)
+func keyClaimedTotal(programName: String) = ["%s%s", "claimedTotal", programName].makeString(separator)
+
+func keyManagerPublicKey() = "%s__managerPublicKey"
+func keyPendingManagerPublicKey() = "%s__pendingManagerPublicKey"
+
+func managerPublicKeyOrUnit() = match keyManagerPublicKey().getString() {
+  case s: String => s.fromBase58String()
+  case _: Unit => unit
+}
+
+func pendingManagerPublicKeyOrUnit() = match keyPendingManagerPublicKey().getString() {
+  case s: String => s.fromBase58String()
+  case _: Unit => unit
+}
+
+func mustManager(i: Invocation) = {
+  let pd = "Permission denied".throw()
+
+  match managerPublicKeyOrUnit() {
+    case pk: ByteVector => i.callerPublicKey == pk || pd
+    case _: Unit => i.caller == this || pd
+  }
+}
+
+@Callable(i)
+func createReferralProgram(programName: String, protramAllowedContract: String) = {
+  ([
+    BooleanEntry(keyProgramName(programName), true),
+    StringEntry(keyProgramAllowedContract(programName), protramAllowedContract)
+  ], unit)
+}
+
+@Callable(i)
+func createPair(
+  programName: String,
+  referrerAddress: String,
+  referralAddress: String,
+  signature: ByteVector
+) = {
+  strict pairReferrerToReferralExistCheck = getString(
+    keyExistsReferrerToReferral(programName, referrerAddress, referralAddress)
+  ) == unit || throw("pair already exists")
+
+  let pair = [programName, referrerAddress, referralAddress].makeString(":").fromBase58String()
```

Comment:
You can't use `fromBase58String` here, can you? Maybe `toBytes` have to be used.